libraries/chain/fork_database.cpp

```diff
@@ -147,109 +146,67 @@ namespace eosio::chain {
+      reset_root_impl( _root );
+
+      unsigned_int size; fc::raw::unpack( ds, size );
+      for( uint32_t i = 0, n = size.value; i < n; ++i ) {
+         bs_t s;
+         fc::raw::unpack( ds, s );
+         // do not populate transaction_metadatas, they will be created as needed in apply_block with appropriate key recovery
+         s.header_exts = s.block->validate_and_extract_header_extensions();
+         add_impl( std::make_shared<bs_t>( std::move( s ) ), mark_valid_t::no, ignore_duplicate_t::no, true, validator );
+      }
+      block_id_type head_id;
+      fc::raw::unpack( ds, head_id );
 
-            auto candidate = index.template get<by_best_branch>().begin();
-            if( candidate == index.template get<by_best_branch>().end() || !bs_accessor_t::is_valid(**candidate) ) {
-               EOS_ASSERT( head->id() == root->id(), fork_database_exception,
-                           "head not set to root despite no better option available; '${filename}' is likely corrupted",
-                           ("filename", fork_db_file) );
-            } else {
-               EOS_ASSERT( !first_preferred( **candidate, *head ), fork_database_exception,
-                           "head not set to best available option available; '${filename}' is likely corrupted",
-                           ("filename", fork_db_file) );
-            }
-         } FC_CAPTURE_AND_RETHROW( (fork_db_file) )
+      if( root->id() == head_id ) {
+         head = root;
+      } else {
+         head = get_block_impl( head_id );
+         EOS_ASSERT( head, fork_database_exception,
+                     "could not find head while reconstructing fork database from file; '${filename}' is likely corrupted",
+                     ("filename", fork_db_file) );
+      }
 
-         std::filesystem::remove( fork_db_file );
+      auto candidate = index.template get<by_best_branch>().begin();
+      if( candidate == index.template get<by_best_branch>().end() || !bs_accessor_t::is_valid(**candidate) ) {
+         EOS_ASSERT( head->id() == root->id(), fork_database_exception,
+                     "head not set to root despite no better option available; '${filename}' is likely corrupted",
+                     ("filename", fork_db_file) );
+      } else {
+         EOS_ASSERT( !first_preferred( **candidate, *head ), fork_database_exception,
+                     "head not set to best available option available; '${filename}' is likely corrupted",
+                     ("filename", fork_db_file) );
       }
    }
 
    template<class BSP>
-   void fork_database_t<BSP>::close(const std::filesystem::path& fork_db_file) {
+   void fork_database_t<BSP>::close(std::ofstream& out) {
       std::lock_guard g( my->mtx );
-      my->close_impl(fork_db_file);
+      my->close_impl(out);
    }
 
    template<class BSP>
-   void fork_database_impl<BSP>::close_impl(const std::filesystem::path& fork_db_file) {
-      if( !root ) {
-         if( index.size() > 0 ) {
```

Comment:
At one point in Savanna a node could be in a state where it had no reversible blocks. Now that we move LIB via block QCs we should always have a few reversible blocks. Likely good to allow an empty index. We will still want to always write out `head` and `root` anyway.

Reply:
The reason I had a member `in_valid_state` originally is that I tried checking not only the `root` but also the `head` initially before saving a fork_db. This caused an issue, making some tests not pass, so I matched what we did before i.e. saving the fok_db even if the head is undefined.

What we did before was maybe even worse because we didn't write anything ito the file if the head was not defined, but on open we still tried to read it anyways.